4-custom-evaluations.ipynb

```diff
@@ -0,0 +1,744 @@
+    "Note that we also return the result of `LLMObs.export_span()` at the end of the `ask_docs` function. We'll need the exported span for later when we submit evaluation results to Datadog."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from ddtrace.llmobs import LLMObs\n",
+    "from ddtrace.llmobs.decorators import retrieval, workflow\n",
+    "\n",
+    "\n",
+    "@retrieval\n",
+    "def retrieve_nodes(query, retriever=base_retriever):\n",
+    "    nodes = retriever.retrieve(query)\n",
+    "    LLMObs.annotate(\n",
+    "        tags={\"retriever\": retriever.__class__.__name__},\n",
+    "        metadata={\"top_k\": TOP_K},\n",
+    "        input_data=query,\n",
+    "        output_data=[\n",
+    "            {\"text\": node.text, \"id\": node.node_id, \"score\": node.score}\n",
+    "            for node in nodes\n",
+    "        ],\n",
+    "    )\n",
+    "    return nodes\n",
+    "\n",
+    "\n",
+    "@workflow\n",
+    "def ask_docs(\n",
+    "    query, retriever=base_retriever, response_synthesizer=response_synthesizer_compact\n",
+    "):\n",
+    "    LLMObs.annotate(input_data=query)\n",
+    "    nodes = retrieve_nodes(query, retriever=retriever)\n",
+    "    response = response_synthesizer.synthesize(query, nodes=nodes)\n",
+    "    LLMObs.annotate(output_data=str(response))\n",
+    "    return response, LLMObs.export_span()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Our RAG workflow is ready! Try a question about LLM Observability. What do you think of the answer quality?"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Answer: Use the steps provided below to run a simple Python script that generates an LLM Observability trace. To begin, ensure you have the necessary prerequisites in place, such as a Datadog API key and an OpenAI API key stored in your environment as `OPENAI_API_KEY`. You will also need to have the OpenAI Python library installed. Once you have these prerequisites set up, you can proceed with running the script as outlined in the quickstart guide.\n",
+      "Context: ['Prerequisites\\n\\n- LLM Observability requires a Datadog API key (see [the instructions for creating an API key][7]).\\n- The example script below uses OpenAI, but you can modify it to use a different provider. To run the script as written, you need:\\n    - An OpenAI API key stored in your environment as `OPENAI_API_KEY`. To create one, see [Account Setup][4] and [Set up your API key][6] in the OpenAI documentation.\\n    - The OpenAI Python library installed. See [Setting up Python][5] in the OpenAI documentation for instructions.', 'Command-line quickstart\\n\\nUse the steps below to run a simple Python script that generates an LLM Observability trace.']\n"
```

Comment:
I think this is ok for now since someone going through the guide is not expected to follow these links (this part of the guide represents the raw text retrieved from our public datadog docs github)